ibis/sql/compiler.py

```diff
@@ -626,42 +620,39 @@ def get_result(self):
         for clause in self.query.filters:
             self.visit(clause)
 
-        to_extract = []
-
-        # Read them inside-out, to avoid nested dependency issues
-        for expr, key in zip(self.observed_exprs.keys,
-                             self.observed_exprs.values):
-            v = self.expr_counts[key]
+        expr_counts = self.expr_counts
 
-            if self.greedy or v > 1:
-                to_extract.append(expr)
+        if self.greedy:
+            to_extract = list(expr_counts.keys())
+        else:
+            to_extract = [op for op, count in expr_counts.items() if count > 1]
 
-        return to_extract
+        node_to_expr = self.node_to_expr
+        return [node_to_expr[op] for op in to_extract]
 
     def observe(self, expr):
-        if self.seen(expr):
-            key = self.observed_exprs.get(expr)
-        else:
-            # this key only needs to be unique because of the IbisMap
-            key = id(expr.op())
-            self.observed_exprs.set(expr, key)
+        key = expr.op()
+
+        if key not in self.node_to_expr:
+            self.node_to_expr[key] = expr
 
-        self.expr_counts[key] += 1
+        assert self.node_to_expr[key].equals(expr)
```

Comment:
Does it test the hashing in the compiler implementation?

Reply:
Yes, this is really just to confirm that `__hash__` matches `equals`, since we can't have `__hash__` match `__eq__`.

Reply:
That's True for expressions, but ops have `__hash__` match `__eq__`. Am I wrong?

Reply:
That is correct. Since `Node`s don't need `__eq__` to return anything but `True` or `False` they can match.

Reply:
I thought that assertion is unnecessary :)

Reply:
But actually there is the `magic`.

Reply:
So the key was to prevent calling `Expr.__eq__` which required to prevent having a collision.

Reply:
However if someone (e.g. me) blindly starts to put Expressions into dicts, then strange things can happen.

Reply:
> So the key was to prevent calling `Expr.__eq__` which required to prevent having a collision.

That is correct. I'm not actually sure we should make `Expr` hashable since we can't override `__eq__`. This means that things like:

```
t = ibis.table([('a', 'int64')])
d = {t.a: 1}
d[t.a]
```
Will raise an exception despite sharing a hash because if two objects hash to the same value but they point to different things then Python will try to call the `__eq__` method to see if the objects compare equal.

Reply:
The only other way to do this would be to ensure that there was only every one instance of every expression with the same hash.